generate_connectors.py

```diff
@@ -150,139 +157,266 @@ def _uuid(identifier: str) -> str:
         y = get_y(pin, per_row, 1, spacing, False)
         top_offset = offset if pin == 1 else 0
         bot_offset = offset if pin == per_row else 0
-        lines.append('   (vertex (position -{} {}) (angle 0.0))'.format(ff(x_inner), ff(y - 1.27 - bot_offset)))
-        lines.append('   (vertex (position -{} {}) (angle 0.0))'.format(ff(x_outer), ff(y - 1 - bot_offset)))
-        lines.append('   (vertex (position -{} {}) (angle 0.0))'.format(ff(x_outer), ff(y + 1 + top_offset)))
+        polygon.add_vertex(Vertex(Position(-x_inner, y - 1.27 - bot_offset), Angle(0)))
+        polygon.add_vertex(Vertex(Position(-x_outer, y - 1 - bot_offset), Angle(0)))
+        polygon.add_vertex(Vertex(Position(-x_outer, y + 1 + top_offset), Angle(0)))
     # Back to start
     top_y = get_y(1, per_row, 1, spacing, False) + spacing / 2 + offset
-    lines.append('   (vertex (position -{} {}) (angle 0.0))'.format(ff(x_inner), ff(top_y)))
-    lines.append('   (vertex (position {} {}) (angle 0.0))'.format(ff(x_inner), ff(top_y)))
-    lines.append('   (vertex (position {} {}) (angle 0.0))'.format(ff(x_outer), ff(top_y - 0.27)))
-    lines.append('  )')
+    polygon.add_vertex(Vertex(Position(-x_inner, top_y), Angle(0)))
+    polygon.add_vertex(Vertex(Position(x_inner, top_y), Angle(0)))
+    polygon.add_vertex(Vertex(Position(x_outer, top_y - 0.27), Angle(0)))
+
+    return polygon
+
+
+def generate_3d_model_generic(
+    model_type: str,  # male or female
+    library: str,
+    full_name: str,
+    uuid_pkg: str,
+    uuid_3d: str,
+    rows: int,
+    pin_count: int,
+    drill: float,
+) -> None:
+    import cadquery as cq
+
+    from cadquery_helpers import StepAssembly, StepColor
+
+    print(f'Generating pkg 3D model "{full_name}": {uuid_3d}')
+
+    insulator_height = 7.5  # Of socket strip
+    standoff_height = 3.0
+    if model_type == 'female':
+        # These are often slightly flat
+        lead_dimensions = (0.4, drill - 0.2)
+    else:
+        dim = math.sqrt(((drill - 0.05) ** 2) / 2)
+        lead_dimensions = (dim, dim)
+    lead_length = 3.0  # Towards bottom side
+
+    # Make base element a little longer, to get some overlap (to avoid rendering bugs when two
+    # faces are in the same position)
+    a_little = 0.001
+
+    # Insulator
+    if model_type == 'female':
+        insulator = cq.Workplane('XY') \
+            .box(spacing, spacing + a_little, insulator_height, centered=(True, True, False)) \
+            .transformed(offset=(0, 0, standoff_height)) \
+            .rect(spacing / 1.5, spacing / 1.5) \
+            .offset2D(spacing / 20) \
+            .cutBlind(insulator_height - standoff_height)
+    else:
+        insulator = cq.Workplane('XY') \
+            .box(spacing, spacing + a_little, standoff_height, centered=(True, True, False))
+
+    # Lead
+    if model_type == 'female':
+        total_lead_length = lead_length
+    else:
+        total_lead_length = lead_length + insulator_height
```

Comment:
@dbrgn Just to confirm, here's actually `standoff_height` missing, right? Or is this by intention? Asking because the pins look very short to me.

```suggestion
        total_lead_length = lead_length + standoff_height + insulator_height
```

(you don't need to fix it, I have to push another fix anyway)

Reply:
Hm with this calculation now it feels a bit long :see_no_evil: 

![image](https://github.com/LibrePCB/librepcb-parts-generator/assets/5374821/1a540b44-8be5-4521-af5d-ca3e39c52d61)

So now I'm a bit unsure what would actually be the intended length of the pins. From the logical point of view, IMHO it would make more sense to include `standoff_height` in the addition to make the exposed male pin length identical to the female hole depth (or very slightly shorter) but I might misunderstand the idea of the calculations.

Reply:
This was my intention:

![image](https://github.com/LibrePCB/librepcb-parts-generator/assets/105168/d04c6f13-91cc-441e-9a11-8275b9019c1a)

The male and female pins should match. However, I made the female socket strip slightly less high than the one I have here. Additionally, I think the offset in the female insulator should probably be less high than for the male headers.

I'll make an adjustment!

Reply:
Thanks for clarifying! I think the main reason why I was confused is that the standoff height is actually used for two unrelated things.

> Additionally, I think the offset in the female insulator should probably be less high than for the male headers.

I agree :+1: I think a pretty small value would be sufficient, and making it independent of the male standoff height makes logically more sense since they are not related to each other.

Reply:
Hmm, actually this system _does_ seem to be correct:

![image](https://github.com/LibrePCB/librepcb-parts-generator/assets/105168/d39e1a30-75dc-4e3b-86f0-540c82e9071a)

However, the insulator height is 8mm, not 7.5mm. (I reduced it slightly because there are many socket strips around on Digikey that are less high.)

This is how it looks with 8mm:

![screenshot-20231021-002801](https://github.com/LibrePCB/librepcb-parts-generator/assets/105168/1ff362ef-d283-4d67-8506-152370a84a6f)

What we could do as well, is to make the insulator height 7mm, and the standoff height 2mm. That saves space without reducing exposed pin height.

I pushed a fixup commit, what do you think?

Reply:
I pushed a second fixup commit, with decoupled config.

Reply:
LGTM, thanks! :+1: